src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            var assemblyLocation = assembly.Location;
+
+            if (string.IsNullOrWhiteSpace(assemblyLocation) && !string.IsNullOrWhiteSpace(assembly.CodeBase))
+            {
+                var uri = new UriBuilder(assembly.CodeBase);
+
+                // Normalize the path to a UNC path. This will remove things like file:// from start of the uri.Path.
+                assemblyLocation = Uri.UnescapeDataString(uri.Path);
+
+                // Still couldn't resolve a valid assemblyLocation.
+                if (string.IsNullOrWhiteSpace(assemblyLocation))
+                {
+                    return null;
+                }
+            }
+
+            var xmlDocumentationFile = GetXmlDocumentationFile(assemblyLocation);
+
+            // We only want to process the file if it exists.
+            if (xmlDocumentationFile != null)
+            {
+                var documentationProvider = new XmlDocumentationProvider(xmlDocumentationFile.FullName);
+
+                var summary = documentationProvider.GetSummary(id);
+                var remarks = documentationProvider.GetRemarks(id);
+
+                if (!string.IsNullOrEmpty(summary) || !string.IsNullOrEmpty(remarks))
+                {
+                    return new TagHelperUsageDescriptor(summary, remarks);
+                }
+            }
+
+            return null;
+        }
+
+        private static FileInfo GetXmlDocumentationFile(string assemblyLocation)
+        {
+            if (string.IsNullOrWhiteSpace(assemblyLocation) ||
+                assemblyLocation.IndexOfAny(InvalidPathCharacters) != -1)
+            {
+                return null;
+            }
+
+            try
+            {
+                var assemblyDirectory = Path.GetDirectoryName(assemblyLocation);
+                var assemblyName = Path.GetFileName(assemblyLocation);
+                var assemblyXmlDocumentationName = Path.ChangeExtension(assemblyName, ".xml");
+                var culture = CultureInfo.CurrentCulture;
+
+                var assemblyXmlDocumentationFile = new FileInfo(
+                    Path.Combine(assemblyDirectory, assemblyXmlDocumentationName));
+
+                // If there's not an XML file side-by-side the .dll it may exist in a culture specific directory.
+                if (!assemblyXmlDocumentationFile.Exists)
+                {
+                    var fallbackDirectories = GetCultureFallbackDirectories();
+                    assemblyXmlDocumentationFile = fallbackDirectories
+                        .Select(fallbackDiretory =>
+                            new FileInfo(
+                                Path.Combine(assemblyDirectory, fallbackDiretory, assemblyXmlDocumentationName)))
+                        .FirstOrDefault(file => file.Exists);
+                }
+
+                return assemblyXmlDocumentationFile;
+            }
+            catch (PathTooLongException)
```

Comment:
:+1:, I added it because it was the only other exception the `Path` utilities could throw (I already handle the invalid path characters). Will remove.